feat: Parse charms as gears, show charm identification percentages

kristofbolyai opened the pull request:
Review by commit. I tried to make it integrate as neatly as I could, but I am aware that some parts are a bit more shoehorned than we would like it to be ideally, but I do believe this is the nicest way to integrate this overall, with minor modifications.

Some screenshots:
Note that charm ids are inverted differently than spell stats. This means they also have 30-130 as the internal roll range, not taking the base value sign into account. They also have the best effects on the lowest internal rolls.
![image](https://github.com/Wynntils/Artemis/assets/49001742/9f9878f8-3708-455f-9237-3822ad7280e1)
![image](https://github.com/Wynntils/Artemis/assets/49001742/b34f86ba-6ac6-4aa0-ab05-d6db9ce34724)
![image](https://github.com/Wynntils/Artemis/assets/49001742/55cc35d5-78d8-48ac-b028-1a3a8d61de9c)
![image](https://github.com/Wynntils/Artemis/assets/49001742/929f3fd3-c377-4e20-9149-a72ae9086c9e)

Other items:
![image](https://github.com/Wynntils/Artemis/assets/49001742/cc7596d0-0c7e-4cea-8d06-4f14dd73c7fb)
![image](https://github.com/Wynntils/Artemis/assets/49001742/edb0aae1-e8f0-4392-a325-b509f1034734)

kristofbolyai committed:
chore: As a prerequisite to parsing charm info, parse tome base stats correctly

Tome base stats are a special kind of identifications. Essentially they behave just like normal identifications, except that they always have a pre-identified, static value, and they are displayed in a non-regular manner. This kind of info is not used at the moment, but they will be used in the tome guide, and when sharing tomes in chat, with the new gear encoding.

Charms have this kind of stats as well, except that they display it like normal stats, and are variable.

See https://i.imgur.com/AD8O0y2.png for an example.

kristofbolyai committed:
chore: Rename CharmInfo to DeprecatedCharmInfo

kristofbolyai committed:
feat: Load charm info from API

kristofbolyai committed:
feat: Make it work with id parsing and ItemStatInfoFeature

kristofbolyai committed:
chore: remove DeprecatedCharmInfo

kristofbolyai committed:
Merge branch 'main' into charm-parsing

kristofbolyai commented on common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java:
```diff
@@ -190,11 +190,11 @@ public static WynnItemParseResult parseItemStack(
                 String statDisplayName = statMatcher.group(6);
 
+                // Skill bonuses looks like stats when parsing, ignore them
+                if (Skill.isSkill(statDisplayName)) continue;
```
This is a low hanging optimization, which I've done to make the performance impact of the regex-based display names lookups close to 0.

kristofbolyai committed:
fix: Fix bug in variable stat deserialization causing ranges to be calculated for items with fixed stats

magicus commented on common/src/main/java/com/wynntils/models/stats/builders/MiscStatKind.java:
```diff
@@ -44,30 +45,50 @@ public enum MiscStatKind {
+
+    // Charm stats
+    LEVELED_XP_BONUS(
+            "XP from Lv. ${} content",
```
So this is more like a "stat template", where you can have different instances with specific levels? That really messes up our system, doesn't it? We'l need to store the level as part of the stat "key", not as the stat value...

kristofbolyai replied:
It only had to be patched in StatModel for getting the display name from string and for getting a stat from display name. Although it is kind of hacky, I am not sure I can see a more obvious way forward atm.

magicus commented:
I'm honestly not sure I understood how you are handling the level-variable stats. I will need to read through the code again when I'm more sharp. Apart from that, it looks fine I guess, it was just more of a "standard" conversion.

kristofbolyai commented on common/src/main/java/com/wynntils/models/stats/StatModel.java:
@magicus This class' changes show everything that needs to be changed to handle leveled stats.

magicus commented:
Okay, so you're using the levelRange as a requirement on the entire charm, and then you inject this when displaying a certain stat for that charm? This seems to be mandated by the API design, I guess. So if Wynn ever introduces more stats that have a level range, that level range will have to be the same for all stats on that particular charm. (Or they will have to redo the API.)

I guess this works.

kristofbolyai committed:
chore: remove unused method

kristofbolyai committed:
refactor: In preparation for charm identifications, add support for custom stat type ranges, rework the inverted stat system, other fixes to internal rolls (#2282)

* refactor: Allow custom base ranges for items, rework inverted calculation, add option to display stat as inverted

* fix: Fix SpellStatType internal names for percentage values

* fix: Fix elemental defense stats

Turns out the API was wrong, and the removed type does not exist, and is only a typo. This is now confirmed to work in-game.

* fix: Correctly calculate internal role in every case

I've done throughout testing, but I think just running with the assertions enabled without any issues proves that this should work.

* feat: Reworked perfect/increase/decrease percentages

* fix: fix legacy order not displaying skill stats (#2283)

* chore: [auto-generated] Update urls.json [ci skip] (#2284)

Co-authored-by: magicus <magicus@users.noreply.github.com>

* feat: Pet menu search (#2285)

* feat: Pet menu search

* ci: spotless formatting

---------

Co-authored-by: DonkeyBlaster <DonkeyBlaster@users.noreply.github.com>

---------

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: magicus <magicus@users.noreply.github.com>
Co-authored-by: Ryan <57310593+DonkeyBlaster@users.noreply.github.com>
Co-authored-by: DonkeyBlaster <DonkeyBlaster@users.noreply.github.com>

kristofbolyai committed:
Merge branch 'main' into charm-parsing

# Conflicts:
#	common/src/main/java/com/wynntils/features/tooltips/ItemStatInfoFeature.java
#	common/src/main/java/com/wynntils/handlers/tooltip/gear/IdentifiableGearItemInfo.java
#	common/src/main/java/com/wynntils/handlers/tooltip/tome/IdentifiableTomeItemInfo.java
#	common/src/main/java/com/wynntils/models/rewards/TomeInfoRegistry.java
#	common/src/main/java/com/wynntils/models/rewards/type/TomeInfo.java
#	common/src/main/java/com/wynntils/models/stats/builders/DefenceStatBuilder.java
#	common/src/main/java/com/wynntils/models/wynnitem/AbstractItemInfoDeserializer.java

kristofbolyai committed:
fix: Properly calculate charm stats

kristofbolyai committed:
fix: fix inconsistent elemental defence key handling

kristofbolyai commented:
@magicus Do you have time to review this, or should I self-review and merge?

kristofbolyai committed:
Merge branch 'main' into charm-parsing

# Conflicts:
#	common/src/main/java/com/wynntils/models/stats/StatCalculator.java
#	common/src/main/java/com/wynntils/models/stats/builders/DefenceStatBuilder.java
#	common/src/main/java/com/wynntils/models/stats/type/StatType.java

kristofbolyai committed:
refactor: rethink how treatAsNegative works

We should not modify the internal roll, as it causes issues in the future, for gear encoding. Also handle star calculations.

kristofbolyai committed:
chore: remove old note in CharmInfo

kristofbolyai committed:
fix: fix a bug with displayAsInverted/treatAsInverted in StatType

kristofbolyai committed:
chore: remove one more old comment

kristofbolyai commented:
I've self-reviewed this, should be good. I will merge this before the announcement on discord.
